File: hard/next_number.hs
import System.Environment (getArgs)

dsig  :: Int -> Int
dsig 0 = 0
dsig x | mod x 10 > 0 = 2^(3 * mod x 10) + dsig (div x 10)
       | otherwise    = dsig (div x 10)

nextNu    :: Int -> Int -> Int
nextNu x y | x == dsig y = y
           | otherwise   = nextNu x (y + 9)

nextNum  :: Int -> Int
nextNum x = nextNu (dsig x) (x + 9)

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map (show . nextNum . read) $ lines input

File: hard/next_number.lua
function div(a, b) return math.floor(a/b) end

function dsig(a)
  local r = 0
  while a > 0 do
    if a%10 > 0 then
      r = r + 2^(3*(a%10))
    end
    a = div(a, 10)
  end
  return r
end

for line in io.lines(arg[1]) do
  local d = tonumber(line)
  local ds, e = dsig(d), d + 9
  while dsig(e) ~= ds do
    e = e + 9
  end
  print(e)
end
